File: 3627-find-minimum-time-to-reach-last-room-i/find-minimum-time-to-reach-last-room-i.java
class Solution {
    // Mảng `d` để hỗ trợ duyệt 4 hướng (trái, phải, lên, xuống)
    static final int[] d = {0, 1, 0, -1, 0};

    // Lớp đại diện cho mỗi ô (cell) trong mê cung, chứa thời gian đến được ô đó và tọa độ (i, j)
    static class Cell implements Comparable<Cell> {
        int time, i, j;

        // Constructor
        Cell(int time, int i, int j) {
            this.time = time;
            this.i = i;
            this.j = j;
        }

        // Sắp xếp các ô theo thời gian tăng dần (để dùng trong PriorityQueue)
        public int compareTo(Cell other) {
            return Integer.compare(this.time, other.time);
        }
    }

    // Hàm chính để tìm thời gian tối thiểu đến đích (n-1, m-1)
    public int minTimeToReach(int[][] moveTime) {
        int n = moveTime.length, m = moveTime[0].length;

        // Khởi tạo mảng lưu thời gian nhỏ nhất để đến từng ô, ban đầu gán là vô cực
        int[][] time = new int[n][m];
        for (int[] row : time) Arrays.fill(row, Integer.MAX_VALUE);

        // PriorityQueue (min-heap) để chọn ô có thời gian nhỏ nhất tiếp theo
        PriorityQueue<Cell> pq = new PriorityQueue<>();

        // Bắt đầu từ ô (0,0) với thời gian 0
        pq.offer(new Cell(0, 0, 0));
        time[0][0] = 0;

        // Bắt đầu duyệt theo thuật toán Dijkstra
        while (!pq.isEmpty()) {
            Cell cur = pq.poll(); // Lấy ô có thời gian nhỏ nhất ra

            // Nếu đã tới đích thì trả về thời gian
            if (cur.i == n - 1 && cur.j == m - 1) {
                return cur.time;
            }

            // Duyệt 4 hướng từ ô hiện tại
            for (int k = 0; k < 4; k++) {
                int ni = cur.i + d[k];     // Hàng kế bên
                int nj = cur.j + d[k + 1]; // Cột kế bên

                // Nếu nằm ngoài biên thì bỏ qua
                if (ni < 0 || nj < 0 || ni >= n || nj >= m) continue;

                // Nếu đến ô (ni, nj) sớm hơn moveTime[ni][nj], thì phải đợi đến đúng thời điểm mới được bước vào
                // Sau đó mất thêm 1 giây để đi
                int waitTime = Math.max(cur.time, moveTime[ni][nj]) + 1;

                // Nếu thời gian mới tốt hơn thời gian đã lưu trước đó
                if (waitTime < time[ni][nj]) {
                    time[ni][nj] = waitTime;        // Cập nhật thời gian mới tốt hơn
                    pq.offer(new Cell(waitTime, ni, nj)); // Đưa ô này vào hàng chờ
                }
            }
        }

        // Không thể đến đích (theo đề bài thì không xảy ra), trả -1 nếu xảy ra
        return -1;
    }
}

File: 3627-find-minimum-time-to-reach-last-room-i/find-minimum-time-to-reach-last-room-i.go
// Cấu trúc phần tử trong priority queue: chứa thời gian, và tọa độ (i, j)
type Cell struct {
	time int;
	i, j int;
}

// PriorityQueue cài đặt heap.Interface
type PriorityQueue []Cell;

func (pq PriorityQueue) Len() int { return len(pq); }

func (pq PriorityQueue) Less(i, j int) bool {
	return pq[i].time < pq[j].time; // Sắp xếp theo thời gian tăng dần
}

func (pq PriorityQueue) Swap(i, j int) { pq[i], pq[j] = pq[j], pq[i]; }

func (pq *PriorityQueue) Push(x any) {
	*pq = append(*pq, x.(Cell));
}

func (pq *PriorityQueue) Pop() any {
	old := *pq;
	n := len(old);
	item := old[n-1];
	*pq = old[:n-1];
	return item;
}

// Hàm chính
func minTimeToReach(moveTime [][]int) int {
	n := len(moveTime);
	m := len(moveTime[0]);

	// Duyệt 4 hướng: phải, xuống, trái, lên
	d := []int{0, 1, 0, -1, 0};

	// Mảng lưu thời gian ngắn nhất đến mỗi ô
	time := make([][]int, n);
	for i := 0; i < n; i++ {
		time[i] = make([]int, m);
		for j := 0; j < m; j++ {
			time[i][j] = math.MaxInt;
		}
	}

	// Priority queue
	pq := &PriorityQueue{};
	heap.Init(pq);
	heap.Push(pq, Cell{0, 0, 0});
	time[0][0] = 0;

	// Bắt đầu Dijkstra
	for pq.Len() > 0 {
		cur := heap.Pop(pq).(Cell);

		// Nếu tới đích thì return
		if cur.i == n-1 && cur.j == m-1 {
			return cur.time;
		}

		for k := 0; k < 4; k++ {
			ni, nj := cur.i+d[k], cur.j+d[k+1];
			if ni < 0 || nj < 0 || ni >= n || nj >= m {
				continue; // Ngoài biên
			}

			// Tính thời gian cần đợi nếu tới sớm hơn moveTime
			wait := max(cur.time, moveTime[ni][nj]) + 1;

			if wait < time[ni][nj] {
				time[ni][nj] = wait;
				heap.Push(pq, Cell{wait, ni, nj});
			}
		}
	}

	return -1; // Không thể đến đích (đề không yêu cầu trường hợp này)
}

// Hàm max
func max(a, b int) int {
	if a > b {
		return a;
	}
	return b;
}